mne/preprocessing/hfc.py

```diff
@@ -0,0 +1,128 @@
+
+    Remove evironmental fields from magentometer data by assuming it is
+    explained as a homogeneous :footcite:`TierneyEtAl2021` or harmonic field
+    :footcite:`TierneyEtAl2022`. Useful for arrays of OPMs.
+
+    Parameters
+    ----------
+    info : instance of Info
+        Info from an instance of Raw (preferable).
+    order : int
+        The order of the spherical harmonic basis set to use. Set to 1 to use
+        only the homogeneous field component (default), 2 to add gradients, 3
+        to add quadrature terms etc.
+    picks : str | array_like | slice | None
+        Channels to include. Default of ``'meg'`` will select all non-reference
+        MEC channels.
+    exclude : list | 'bads'
+        List of channels to exclude from HFC, only used when picking
+        based on types (e.g., exclude="bads" when picks="meg").
+        Specify ``'bads'`` (the default) to exclude all channels marked as bad.
+    ignore_ref : bool
+        Specify whether reference MEG channels should be ignored when
+        calculating the basis set (default: ``True``).
+    accuracy : str
+        Can be ``"point"``, ``"normal"`` or ``"accurate"`` (default), defines
+        which level of coil definition accuracy is used to generate model.
+    %(verbose)s
+
+    Returns
+    -------
+    projs : list of Projection
+        List of projection vectors.
+
+    Notes
+    -----
+    To apply the projectors to a dataset, use ``inst.add_proj(projs).apply
+    _proj()``.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    idx = _picks_to_idx(info, picks, exclude=exclude)
+    if not ignore_ref:
+        idx = np.union1d(idx,
+                         _picks_to_idx(info, picks='ref_meg', exclude=exclude))
+    basis, channels = _generate_basis_set(info, idx, order,
+                                          accuracy, ignore_ref)
+    labels = _label_basis(order)
+    _assert_isfinite(basis, channels)
+    projs = []
+    for ii, label in enumerate(labels):
+        data = basis[:, ii]
+        proj_data = dict(col_names=channels, row_names=None,
+                         data=data[np.newaxis, :], ncol=len(channels), nrow=1)
+        proj = Projection(active=False, data=proj_data, desc=label)
+        projs.append(proj)
+
+    return projs
+
+
+def _generate_basis_set(info, picks=None, order=1, accuracy='accurate',
+                        ignore_ref=True, origin=(0, 0, 0)):
+    """Generate the basis set used for HFC."""
+    exp = dict(origin=origin, int_order=0, ext_order=order)
+    mf_coils = _prep_mf_coils(info, ignore_ref=ignore_ref,
+                              accuracy=accuracy, return_names=True)
+    coils = mf_coils[:-1]
+    mf_names = mf_coils[-1]
+    S = _sss_basis(exp, coils)
+    if picks is not None:
+        pick_names = [info['chs'][i]['ch_name'] for i in picks]
+        basis_chans, basis_picks = _filter_list(pick_names, mf_names)
+        S = S[basis_picks]
```

Comment:
Okay I looked closer and now I see what's going on that leads to the need to `_filter_list` -- if you only use a subset of channels via `picks` you might as well only select *those channels* before "prepping" them. In other words, instead of the current code, you can do
```suggestion
                        *, origin=(0, 0, 0)):
    """Generate the basis set used for HFC."""
    exp = dict(origin=origin, int_order=0, ext_order=order)
    info = pick_info(info, picks)
    coils = _prep_mf_coils(info, ignore_ref=False, accuracy=accuracy)
    S = _sss_basis(exp, coils)
```
then the `coils` you get here should be the correct set, no need for `_filter_names` anymore. And no need for the `return_names`, either, I think.

I also added `ignore_ref=False` here (and no longer take it as an ) because subselection to the correct channels should be taken care of by `_picks_to_idx` already (no need to deselect any channels anymore).